### ISSUE TITLE
dbal tracing does not work with PdoSessionHandler

### ISSUE THREAD

**Issue by retnek** [getsentry/sentry-symfony#603]
### Environment

How do you use Sentry?
Sentry SaaS (sentry.io)

Which SDK and version?
sentry/sentry-symfony 4.2.6
symfony version 4.4
php 7.3
MacOsx

### Steps to Reproduce

1. set session handler mode in framework.yaml to as described here: https://symfony.com/doc/4.4/session/database.html#store-sessions-in-a-relational-database-mariadb-mysql-postgresql
2. setup sentry as described here: https://docs.sentry.io/platforms/php/guides/symfony/
3. launch symfony app with php-fpm using `symfony serve` command, then you got Uncaught Error: Argument 1 passed to Symfony\Component\HttpFoundation\Session\Storage\Handler\PdoSessionHandler::connect() must be of the type string, object given, called in /some/where/vendor/symfony/http-foundation/Session/Storage/Handler/PdoSessionHandler.php on line 271.
The `PdoSessionHandler::connect` waits `$this->dsn` as string, but in this case its an object:
`Sentry\SentryBundle\Tracing\Doctrine\DBAL\TracingServerInfoAwareDriverConnection`

### Expected Result

Our symfony application starts as before we install sentry.io.

### Actual Result

We got a critical exception:

```
request.CRITICAL: Uncaught PHP Exception TypeError: "Argument 1 passed to Symfony\Component\HttpFoundation\Session\Storage\Handler\PdoSessionHandler::connect() must be of the type string, object given, called in /some/where/vendor/symfony/http-foundation/Session/Storage/Handler/PdoSessionHandler.php on line 271" at /some/where/vendor/symfony/http-foundation/Session/Storage/Handler/PdoSessionHandler.php line 441 {"exception":"[object] (TypeError(code: 0): Argument 1 passed to Symfony\\Component\\HttpFoundation\\Session\\Storage\\Handler\\PdoSessionHandler::connect() must be of the type string, object given, called in /some/where/vendor/symfony/http-foundation/Session/Storage/Handler/PdoSessionHandler.php on line 271 at /some/where/vendor/symfony/http-foundation/Session/Storage/Handler/PdoSessionHandler.php:441)
[stacktrace]
#0 /some/where/vendor/symfony/http-foundation/Session/Storage/Handler/PdoSessionHandler.php(271): Symfony\\Component\\HttpFoundation\\Session\\Storage\\Handler\\PdoSessionHandler->connect(Object(Sentry\\SentryBundle\\Tracing\\Doctrine\\DBAL\\TracingServerInfoAwareDriverConnection))
#1 /some/where/vendor/symfony/http-foundation/Session/Storage/Proxy/SessionHandlerProxy.php(44): Symfony\\Component\\HttpFoundation\\Session\\Storage\\Handler\\PdoSessionHandler->open('', 'PHPSESSID')
#2 [internal function]: Symfony\\Component\\HttpFoundation\\Session\\Storage\\Proxy\\SessionHandlerProxy->open('', 'PHPSESSID')
#3 /some/where/vendor/symfony/http-foundation/Session/Storage/NativeSessionStorage.php(156): session_start()
#4 /some/where/vendor/symfony/http-foundation/Session/Storage/NativeSessionStorage.php(323): Symfony\\Component\\HttpFoundation\\Session\\Storage\\NativeSessionStorage->start()
#5 /some/where/vendor/symfony/http-foundation/Session/Session.php(251): Symfony\\Component\\HttpFoundation\\Session\\Storage\\NativeSessionStorage->getBag('attributes')
#6 /some/where/vendor/symfony/http-foundation/Session/Session.php(273): Symfony\\Component\\HttpFoundation\\Session\\Session->getBag('attributes')
#7 /some/where/vendor/symfony/http-foundation/Session/Session.php(73): Symfony\\Component\\HttpFoundation\\Session\\Session->getAttributeBag()
#8 /some/where/vendor/symfony/security/Http/Firewall/ContextListener.php(119): Symfony\\Component\\HttpFoundation\\Session\\Session->get('_security_main')
#9 /some/where/vendor/symfony/security-bundle/Debug/WrappedLazyListener.php(49): Symfony\\Component\\Security\\Http\\Firewall\\ContextListener->authenticate(Object(Symfony\\Component\\HttpKernel\\Event\\RequestEvent))
#10 /some/where/vendor/symfony/security/Http/Firewall/AbstractListener.php(27): Symfony\\Bundle\\SecurityBundle\\Debug\\WrappedLazyListener->authenticate(Object(Symfony\\Component\\HttpKernel\\Event\\RequestEvent))
#11 /some/where/vendor/symfony/security-bundle/Debug/TraceableFirewallListener.php(62): Symfony\\Component\\Security\\Http\\Firewall\\AbstractListener->__invoke(Object(Symfony\\Component\\HttpKernel\\Event\\RequestEvent))
#12 /some/where/vendor/symfony/security/Http/Firewall.php(98): Symfony\\Bundle\\SecurityBundle\\Debug\\TraceableFirewallListener->callListeners(Object(Symfony\\Component\\HttpKernel\\Event\\RequestEvent), Object(Generator))
#13 /some/where/vendor/symfony/event-dispatcher/Debug/WrappedListener.php(126): Symfony\\Component\\Security\\Http\\Firewall->onKernelRequest(Object(Symfony\\Component\\HttpKernel\\Event\\RequestEvent), 'kernel.request', Object(Symfony\\Component\\HttpKernel\\Debug\\TraceableEventDispatcher))
#14 /some/where/vendor/symfony/event-dispatcher/EventDispatcher.php(264): Symfony\\Component\\EventDispatcher\\Debug\\WrappedListener->__invoke(Object(Symfony\\Component\\HttpKernel\\Event\\RequestEvent), 'kernel.request', Object(Symfony\\Component\\HttpKernel\\Debug\\TraceableEventDispatcher))
#15 /some/where/vendor/symfony/event-dispatcher/EventDispatcher.php(239): Symfony\\Component\\EventDispatcher\\EventDispatcher->doDispatch(Array, 'kernel.request', Object(Symfony\\Component\\HttpKernel\\Event\\RequestEvent))
#16 /some/where/vendor/symfony/event-dispatcher/EventDispatcher.php(73): Symfony\\Component\\EventDispatcher\\EventDispatcher->callListeners(Array, 'kernel.request', Object(Symfony\\Component\\HttpKernel\\Event\\RequestEvent))
#17 /some/where/vendor/symfony/event-dispatcher/Debug/TraceableEventDispatcher.php(168): Symfony\\Component\\EventDispatcher\\EventDispatcher->dispatch(Object(Symfony\\Component\\HttpKernel\\Event\\RequestEvent), 'kernel.request')
#18 /some/where/vendor/symfony/http-kernel/HttpKernel.php(134): Symfony\\Component\\EventDispatcher\\Debug\\TraceableEventDispatcher->dispatch(Object(Symfony\\Component\\HttpKernel\\Event\\RequestEvent), 'kernel.request')
#19 /some/where/vendor/symfony/http-kernel/HttpKernel.php(80): Symfony\\Component\\HttpKernel\\HttpKernel->handleRaw(Object(Symfony\\Component\\HttpFoundation\\Request), 1)
#20 /some/where/vendor/symfony/http-kernel/Kernel.php(201): Symfony\\Component\\HttpKernel\\HttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#21 /some/where/public/index.php(25): Symfony\\Component\\HttpKernel\\Kernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request))
#22 {main}
"} []
```

### Temporary workaround

If we explicit disable dbal tracing in `sentry.yaml`
```
  tracing:
    dbal:
      enabled: false
```
the issue does not shows up anymore.

**Comment by retnek** [comment]
Okay, after I drove deeper into this issue, I saw something interesting in our `services.yaml` file:
```
app.session_handler:
        class: Symfony\Component\HttpFoundation\Session\Storage\Handler\PdoSessionHandler
        arguments:
             - !service { class: PDO, factory: ['@database_connection', 'getWrappedConnection'] }
```
The interesting part here is using PDO class instead of dsn string (or explicit connection parameters), which built by Doctrine like `getWrappedConnection`. This was completely came from Symfony 4.0 documentation, but above 4.4 isn't mentioned anymore. Maybe not supported above 4.4 I don't know. After we dropped that line, and explicitly set database connection - which is seems to be the supported format - the issue is gone.

Long story short: do not use Doctrine like `getWrappedConnection` to build connection to PdoSessionHandler or any other service which using database, if you are using Sentry dbal tracing feature.

**Comment by ste93cry** [comment]
If that's your service definition, it makes sense that is breaks when the tracing feature is enabled because the feature wraps the DBAL connection which in turn wraps the PDO connection. At the begininng all was working because `getWrappedConnection()` was returning the innermost connection, but if you have tracing enabled it returns the Doctrine connection because you have one more nesting level. An easy fix would be to call the method until it doesn't exist anymore on the object: this way you should be able to reach the innermost connection in all cases

**Comment by Jean85** [comment]
`getNativeConnection()` is also deprecated, you should use `getNativeConnection()` that already does what @ste93cry was suggesting: https://github.com/doctrine/dbal/blob/3.3.x/UPGRADE.md#deprecated-connectiongetwrappedconnection-connectionconnect-made-internal

**Comment by ste93cry** [comment]
Given that a workaround has been suggested and an improvement to get the innermost connection will be available with the next minor version, can I close this or do you need more help @retnek?

**Comment by retnek** [comment]
It's working without getNativeConnection right now @ste93cry, thanks!
